applications/credhub-api/src/main/resources/db/migration/mysql/V57_2__delete_related_encrypted_value_when_credential_is_deleted.sql

```diff
@@ -0,0 +1,50 @@
+-- Fix for https://github.com/cloudfoundry/credhub/issues/231
+-- Create delete trigger on credential table instead of the tables directly
+-- reference encrypted_value table because the trigger does not work for
+-- cascade-deleted records in MySQL.
+DROP TRIGGER IF EXISTS tr_credential_deleted;
+DELIMITER |
+CREATE TRIGGER tr_credential_deleted BEFORE DELETE ON credential
+FOR EACH ROW
+BEGIN
+  DROP TEMPORARY TABLE IF EXISTS
+    tmp_credential_version_uuids, tmp_encrypted_value_uuids;
+
+  -- Remember the uuids of the credential_version records being deleted
+  CREATE TEMPORARY TABLE tmp_credential_version_uuids
+    SELECT uuid FROM credential_version
+      WHERE credential_uuid = OLD.uuid;
+
+  -- Remember the uuids of the encrypted_value records to delete
+  CREATE TEMPORARY TABLE tmp_encrypted_value_uuids
+    SELECT encrypted_value_uuid FROM credential_version
+      WHERE credential_uuid = OLD.uuid;
+  INSERT INTO tmp_encrypted_value_uuids (encrypted_value_uuid)
+    SELECT password_parameters_uuid from password_credential
+      WHERE password_credential.uuid IN
+        (SELECT uuid FROM tmp_credential_version_uuids);
+  INSERT INTO tmp_encrypted_value_uuids (encrypted_value_uuid)
+    SELECT password_parameters_uuid from user_credential
+      WHERE user_credential.uuid IN
+        (SELECT uuid FROM tmp_credential_version_uuids);
+
+  -- Set FKs to encrypted_value to null so the encrypted_value
+  -- record can be deleted without violating the FK constraint
+  UPDATE credential_version SET encrypted_value_uuid = NULL
+    WHERE credential_uuid = OLD.uuid;
+  UPDATE password_credential SET password_parameters_uuid = NULL
+    WHERE password_credential.uuid IN
+      (SELECT uuid FROM tmp_credential_version_uuids);
+  UPDATE user_credential SET password_parameters_uuid = NULL
+    WHERE user_credential.uuid IN
+      (SELECT uuid FROM tmp_credential_version_uuids);
+
+  -- Delete the encrypted_value records
+  DELETE FROM encrypted_value WHERE encrypted_value.uuid IN
```

Comment:
Why is a temporary table needed instead of a nested subquery?

Additionally, what is the FK providing since this is bypassing it?

Finally, This appears to be more complex and creating a higher maintenance cost than utilizing then ORMs orphanRemoval capability, how is this being compared to that approach?

Reply:
> Why is a temporary table needed instead of a nested subquery?

Mainly because the values need to be retrieved before running the next query.

> Additionally, what is the FK providing since this is bypassing it?
Not sure what you mean. I do not think that you understood the context. The trigger deletes the records that cannot be handled by FK cascade delete because the FK direction is the other way.

> This appears to be more complex and creating a higher maintenance cost
There is no higher  maintenance cost. The script is written once and done.

> than utilizing then ORMs orphanRemoval capability, how is this being compared to that approach
My understanding is that more than a week was spent in that effort without producing any useful result. Based on what I see, going that route required very involved Java/Kotlin code changes. This requires no Java/Kotlin code change.

Reply:
I am going to rewrite it not to use temp tables. There are potential issues with using temp tables inside trigger as we do not have complete control of how/when the trigger is executed.